### PR TITLE
bugfix: fix the error of active refresh failure of cross-database table metadata

### LINKED ISSUE #6685: get table meta error
- [x] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
seata 1.6.1（看了1.8和2.0的seata代码也有这个问题）
seata 自动刷新 TableMetaCache 时报错 “get table meta error”

### Ⅱ. Describe what happened
jdbc:mysql://x.x.x.x:3306/A
数据库链接字串中配置的数据为A，此时有一个标注@GlobalTransaction的方法被调用，该方法操作了B数据库的table_1,即B.table_1,插入数据.
这时 io.seata.rm.datasource.sql.struct.cache.AbstractTableMetaCache#getTableMeta 虽然正常运行 ，等到当io.seata.rm.datasource.sql.struct.cache.AbstractTableMetaCache#refresh 刷新时 就会报错了。
 “get table meta error”
```
io.seata.rm.datasource.sql.struct.cache.AbstractTableMetaCache-tableMetaChecker_1_1-get table meta error:Table &#39;qmgr.qo_ord_anomalous&#39; doesn&#39;t exist-N/A-java.sql.SQLSyntaxErrorException: Table 'A.table_1' doesn't exist
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:120)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
	at com.mysql.cj.jdbc.StatementImpl.executeQuery$original$sEQlPt7U(StatementImpl.java:1200)
	at com.mysql.cj.jdbc.StatementImpl.executeQuery$original$sEQlPt7U$accessor$fo53R1GS(StatementImpl.java)
	at com.mysql.cj.jdbc.StatementImpl$auxiliary$PTwarWz7.call(Unknown Source)
	at org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstMethodsInter.intercept(InstMethodsInter.java:86)
	at com.mysql.cj.jdbc.StatementImpl.executeQuery(StatementImpl.java)
	at com.alibaba.druid.pool.DruidPooledStatement.executeQuery(DruidPooledStatement.java:296)
	at io.seata.rm.datasource.sql.struct.cache.MysqlTableMetaCache.fetchSchema(MysqlTableMetaCache.java:82)
	at io.seata.rm.datasource.sql.struct.cache.AbstractTableMetaCache.refresh(AbstractTableMetaCache.java:82)
```

### 最终排查因
直接原因：
1.4.2 代码是没有问题，是因为：client.rm.tableMetaCheckEnable 这个值默认为false，不会开启自动刷新cache
1.6.x client.rm.tableMetaCheckEnable（io.seata.rm.datasource.DataSourceProxy#ENABLE_TABLE_META_CHECKER_ENABLE） 默认值就变成了 true 

根本原因：
io.seata.rm.datasource.sql.struct.cache.AbstractTableMetaCache#getTableMeta
   public TableMeta getTableMeta(final Connection connection, final String tableName, String resourceId) {
        if (StringUtils.isNullOrEmpty(tableName)) {
            throw new IllegalArgumentException("TableMeta cannot be fetched without tableName");
        }

        TableMeta tmeta;
        final String key = getCacheKey(connection, tableName, resourceId);
        tmeta = TABLE_META_CACHE.get(key, mappingFunction -> {
            try {
                return fetchSchema(connection, tableName);
            } catch (SQLException e) {
                LOGGER.error("get table meta of the table `{}` error: {}", tableName, e.getMessage(), e);
                return null;
            }
        });

        if (tmeta == null) {
            throw new ShouldNeverHappenException(String.format("[xid:%s]get table meta failed," +
                " please check whether the table `%s` exists.", RootContext.getXID(), tableName));
        }
       **tableName 参数为 B.table_1，而缓存中 tmeta.tableName的值变成了table_1**
        return tmeta;
    }

    public void refresh(final Connection connection, String resourceId) {
        ConcurrentMap<String, TableMeta> tableMetaMap = TABLE_META_CACHE.asMap();
        for (Map.Entry<String, TableMeta> entry : tableMetaMap.entrySet()) {
            String key = getCacheKey(connection, entry.getValue().getTableName(), resourceId);
            if (entry.getKey().equals(key)) {
                try {
 **这里就取到了没有库前缀的tableName：table_1，在链接字符串 jdbc:mysql://x.x.x.x:3306/A 下访问 table_1 于是就报 A.table_1 不能存在 **
                    TableMeta tableMeta = fetchSchema(connection, entry.getValue().getTableName());
                    if (!tableMeta.equals(entry.getValue())) {
                        TABLE_META_CACHE.put(entry.getKey(), tableMeta);
                        LOGGER.info("table meta change was found, update table meta cache automatically.");
                    }
                } catch (SQLException e) {
                    LOGGER.error("get table meta error:{}", e.getMessage(), e);
                }
            }
        }
    }

### 最后
可能seata没有考虑 在同一个程序中跨库访问，但是实际开发过程中由于各种原因会出现这种情况。
在seata团队没有修复之前，如果有这个不太要紧的错误，看着太烦可以设置client.rm.tableMetaCheckEnable = false。

希望可以帮助到一些人。

### PR DESCRIPTION
**Pull request by GoodBoyCoder**
- [x] I have registered the PR [changes](../changes).

### Ⅰ. Describe what this PR did
Fixed the error that may cause failure to refresh table metadata cache across databases under the same resourceId due to lack of original table name information

### Ⅱ. Does this pull request fix one issue?

fixes #6685

### CODE REVIEW

File: rm-datasource/src/main/java/org/apache/seata/rm/datasource/sql/struct/cache/PostgresqlTableMetaCache.java [hunk 1]
```diff
@@ -185,8 +184,9 @@ private TableMeta resultSetMetaToSchema(Connection connection, String tableName)
             }
 
             while (rsTable.next()) {
-                String rsTableName = rsTable.getString("TABLE_NAME");
+                String rsTableCat = rsTable.getString("TABLE_CAT");
```
Comment (funky-eyes) [inline rm-datasource/src/main/java/org/apache/seata/rm/datasource/sql/struct/cache/PostgresqlTableMetaCache.java]: 为什么读取TABLE_CAT
Why read TABLE_CAT

File: rm-datasource/src/main/java/org/apache/seata/rm/datasource/sql/struct/cache/SqlServerTableMetaCache.java [hunk 2]
```diff
@@ -189,8 +186,9 @@ private TableMeta resultSetMetaToSchema(Connection connection, String tableName)
             }
 
             while (rsTable.next()) {
-                String rsTableName = rsTable.getString("TABLE_NAME");
+                String rsTableCat = rsTable.getString("TABLE_CAT");
```
Comment (funky-eyes) [inline rm-datasource/src/main/java/org/apache/seata/rm/datasource/sql/struct/cache/SqlServerTableMetaCache.java]: Why read TABLE_CAT

Comment (GoodBoyCoder) [reply rm-datasource/src/main/java/org/apache/seata/rm/datasource/sql/struct/cache/SqlServerTableMetaCache.java]: 这是废弃方案漏删除的，原本打算存储表名的所有信息，但是后面发现不是很满足要求并且还有点冗余

This is what was omitted in the abandoned solution. I originally planned to store all the information of the table name, but later found that it did not meet the requirements and was a bit redundant.

**Review by funky-eyes — APPROVED** [review]
LGTM

File: rm-datasource/src/main/java/org/apache/seata/rm/datasource/sql/struct/cache/MysqlTableMetaCache.java [hunk 3]
```diff
@@ -48,9 +48,9 @@ public class MysqlTableMetaCache extends AbstractTableMetaCache {
     protected String getCacheKey(Connection connection, String tableName, String resourceId) {
         StringBuilder cacheKey = new StringBuilder(resourceId);
         cacheKey.append(".");
-        //remove single quote and separate it to catalogName and tableName
-        String[] tableNameWithCatalog = tableName.replace("`", "").split("\\.");
-        String defaultTableName = tableNameWithCatalog.length > 1 ? tableNameWithCatalog[1] : tableNameWithCatalog[0];
+        //original: remove single quote and separate it to catalogName and tableName
+        //now: Use the original table name to avoid cache errors of tables with the same name across databases
+        String defaultTableName = tableName.replace("`", "");
```
Comment (jsbxyyx) [inline rm-datasource/src/main/java/org/apache/seata/rm/datasource/sql/struct/cache/MysqlTableMetaCache.java]: mysql有多种转义符号，参考ColumnUtils

Mysql has a variety of escape symbols, refer to ColumnUtils

File: rm-datasource/src/main/java/org/apache/seata/rm/datasource/sql/struct/cache/OracleTableMetaCache.java [hunk 4]
```diff
@@ -98,6 +99,7 @@ protected TableMeta resultSetMetaToSchema(DatabaseMetaData dbmd, String tableNam
 
         try (ResultSet rsColumns = dbmd.getColumns("", schemaName, tableName, "%");
              ResultSet rsIndex = dbmd.getIndexInfo(null, schemaName, tableName, false, true);
+             ResultSet rsTable = dbmd.getTables("", schemaName, tableName, new String[]{"TABLE"});
```
Comment (jsbxyyx) [inline rm-datasource/src/main/java/org/apache/seata/rm/datasource/sql/struct/cache/OracleTableMetaCache.java]: 这行代码是不是没用

Is this line of code useless?

File: rm-datasource/src/main/java/org/apache/seata/rm/datasource/sql/struct/cache/MysqlTableMetaCache.java (hunk 3 above)
Comment (GoodBoyCoder) [reply rm-datasource/src/main/java/org/apache/seata/rm/datasource/sql/struct/cache/MysqlTableMetaCache.java]: Done

File: rm-datasource/src/main/java/org/apache/seata/rm/datasource/sql/struct/cache/OracleTableMetaCache.java (hunk 4 above)
Comment (GoodBoyCoder) [reply rm-datasource/src/main/java/org/apache/seata/rm/datasource/sql/struct/cache/OracleTableMetaCache.java]: Done

**Review by jsbxyyx — APPROVED** [review]
LGTM

**Pull request merged**